# 1. Install Spark and PySpark

In [1]:
import pyspark
from pyspark.sql import SparkSession

spark = SparkSession.builder \
     .master("local[*]") \
     .appName("homework") \
     .getOrCreate()

Setting default log level to "WARN".
To adjust logging level use sc.setLogLevel(newLevel). For SparkR, use setLogLevel(newLevel).


24/03/07 16:52:22 WARN NativeCodeLoader: Unable to load native-hadoop library for your platform... using builtin-java classes where applicable
24/03/07 16:52:24 WARN Utils: Service 'SparkUI' could not bind on port 4040. Attempting port 4041.


In [2]:
spark.version

'3.3.2'

The output is '3.3.2'

# 2. Read FHV 2019-10
## step 1 repartition(6)
## step2 write.parquet

In [3]:
!wget -O data/fhv/fhv_reipdata_2019-10.csv.gz https://github.com/DataTalksClub/nyc-tlc-data/releases/download/fhv/fhv_tripdata_2019-10.csv.gz 

--2024-03-06 09:23:38--  https://github.com/DataTalksClub/nyc-tlc-data/releases/download/fhv/fhv_tripdata_2019-10.csv.gz
Resolving github.com (github.com)... 140.82.121.4
Connecting to github.com (github.com)|140.82.121.4|:443... connected.
HTTP request sent, awaiting response... 302 Found
Location: https://objects.githubusercontent.com/github-production-release-asset-2e65be/513814948/efdfcf82-6d5c-44d1-a138-4e8ea3c3a3b6?X-Amz-Algorithm=AWS4-HMAC-SHA256&X-Amz-Credential=AKIAVCODYLSA53PQK4ZA%2F20240306%2Fus-east-1%2Fs3%2Faws4_request&X-Amz-Date=20240306T092338Z&X-Amz-Expires=300&X-Amz-Signature=4b04898e920e139da019a08f5d1f0293f4b509018241bcf00249463fe18d37e6&X-Amz-SignedHeaders=host&actor_id=0&key_id=0&repo_id=513814948&response-content-disposition=attachment%3B%20filename%3Dfhv_tripdata_2019-10.csv.gz&response-content-type=application%2Foctet-stream [following]
--2024-03-06 09:23:38--  https://objects.githubusercontent.com/github-production-release-asset-2e65be/513814948/efdfcf82-6d5c-

In [2]:
df_fhv = spark.read \
    .csv("data/fhv/fhv_reipdata_2019-10.csv.gz", header=True, inferSchema=True)

In [19]:
df_fhv.show()

+--------------------+-------------------+-------------------+------------+------------+-------+----------------------+
|dispatching_base_num|    pickup_datetime|   dropOff_datetime|PUlocationID|DOlocationID|SR_Flag|Affiliated_base_number|
+--------------------+-------------------+-------------------+------------+------------+-------+----------------------+
|              B00009|2019-10-01 00:23:00|2019-10-01 00:35:00|         264|         264|   null|                B00009|
|              B00013|2019-10-01 00:11:29|2019-10-01 00:13:22|         264|         264|   null|                B00013|
|              B00014|2019-10-01 00:11:43|2019-10-01 00:37:20|         264|         264|   null|                B00014|
|              B00014|2019-10-01 00:56:29|2019-10-01 00:57:47|         264|         264|   null|                B00014|
|              B00014|2019-10-01 00:23:09|2019-10-01 00:28:27|         264|         264|   null|                B00014|
|     B00021         |2019-10-01 00:00:4

In [20]:
df_fhv.printSchema()

root
 |-- dispatching_base_num: string (nullable = true)
 |-- pickup_datetime: timestamp (nullable = true)
 |-- dropOff_datetime: timestamp (nullable = true)
 |-- PUlocationID: integer (nullable = true)
 |-- DOlocationID: integer (nullable = true)
 |-- SR_Flag: string (nullable = true)
 |-- Affiliated_base_number: string (nullable = true)



In [7]:
df_fhv.repartition(6)

DataFrame[dispatching_base_num: string, pickup_datetime: string, dropOff_datetime: string, PUlocationID: string, DOlocationID: string, SR_Flag: string, Affiliated_base_number: string]

In [21]:
df_fhv.repartition(6).write.parquet("data/fhv/pq", mode="overwrite")

In [22]:
!ls -lh data/fhv/pq

total 39M
-rw-r--r-- 1 xiangivyli xiangivyli    0 Mar  6 09:43 _SUCCESS
-rw-r--r-- 1 xiangivyli xiangivyli 6.4M Mar  6 09:43 part-00000-3cfd9cae-7b01-4f7c-a42e-76b41f6c11e5-c000.snappy.parquet
-rw-r--r-- 1 xiangivyli xiangivyli 6.4M Mar  6 09:43 part-00001-3cfd9cae-7b01-4f7c-a42e-76b41f6c11e5-c000.snappy.parquet
-rw-r--r-- 1 xiangivyli xiangivyli 6.4M Mar  6 09:43 part-00002-3cfd9cae-7b01-4f7c-a42e-76b41f6c11e5-c000.snappy.parquet
-rw-r--r-- 1 xiangivyli xiangivyli 6.4M Mar  6 09:43 part-00003-3cfd9cae-7b01-4f7c-a42e-76b41f6c11e5-c000.snappy.parquet
-rw-r--r-- 1 xiangivyli xiangivyli 6.4M Mar  6 09:43 part-00004-3cfd9cae-7b01-4f7c-a42e-76b41f6c11e5-c000.snappy.parquet
-rw-r--r-- 1 xiangivyli xiangivyli 6.4M Mar  6 09:43 part-00005-3cfd9cae-7b01-4f7c-a42e-76b41f6c11e5-c000.snappy.parquet


The average size of the Parquet is around 6 MB

# 3. 2019-10-15 count the number of records

In [3]:
df_fhv = spark.read.parquet("data/fhv/pq/*")

In [3]:
df_fhv.printSchema()

root
 |-- dispatching_base_num: string (nullable = true)
 |-- pickup_datetime: timestamp (nullable = true)
 |-- dropOff_datetime: timestamp (nullable = true)
 |-- PUlocationID: integer (nullable = true)
 |-- DOlocationID: integer (nullable = true)
 |-- SR_Flag: string (nullable = true)
 |-- Affiliated_base_number: string (nullable = true)



In [4]:
df_fhv.registerTempTable("fhv")

/home/xiangivyli/spark/spark-3.3.2-bin-hadoop3/python/pyspark/sql/dataframe.py:229: FutureWarning: Deprecated in 2.0, use createOrReplaceTempView instead.
  warnings.warn("Deprecated in 2.0, use createOrReplaceTempView instead.", FutureWarning)


In [26]:
df_fhv_count = spark.sql("""
SELECT 
    COUNT(1) AS number
FROM 
    fhv
WHERE 
    pickup_datetime >= "2019-10-15 00:00:00"
AND 
    pickup_datetime < "2019-10-16 00:00:00"
""")

In [27]:
df_fhv_count.show()

+------+
|number|
+------+
| 62610|
+------+



The record is 62,610

# 4. Longest trip for each day

In [36]:
spark.sql("DESCRIBE fhv").show()

+--------------------+---------+-------+
|            col_name|data_type|comment|
+--------------------+---------+-------+
|dispatching_base_num|   string|   null|
|     pickup_datetime|timestamp|   null|
|    dropOff_datetime|timestamp|   null|
|        PUlocationID|      int|   null|
|        DOlocationID|      int|   null|
|             SR_Flag|   string|   null|
|Affiliated_base_n...|   string|   null|
+--------------------+---------+-------+



In [5]:
df_fhv_hour = spark.sql("""
SELECT
    ((unix_timestamp(dropOff_datetime) - unix_timestamp(pickup_datetime)) / 3600) AS trip_hours,
    pickup_datetime,
    dropOff_datetime
FROM 
    fhv
WHERE
    pickup_datetime >= "2019-01-01 00:00:00"
ORDER BY
    trip_hours DESC
""")

In [6]:
df_fhv_hour.show(1000)

+------------------+-------------------+-------------------+
|        trip_hours|    pickup_datetime|   dropOff_datetime|
+------------------+-------------------+-------------------+
|          631152.5|2019-10-11 18:00:00|2091-10-11 18:30:00|
|          631152.5|2019-10-28 09:00:00|2091-10-28 09:30:00|
| 87672.44083333333|2019-10-31 23:46:33|2029-11-01 00:13:00|
| 70128.02805555555|2019-10-01 21:43:42|2027-10-01 21:45:23|
|            8794.0|2019-10-17 14:00:00|2020-10-18 00:00:00|
| 8784.166666666666|2019-10-26 21:26:00|2020-10-26 21:36:00|
|1464.5344444444445|2019-10-30 12:30:04|2019-12-30 13:02:08|
|1056.8266666666666|2019-10-25 07:04:57|2019-12-08 07:54:33|
|1056.2705555555556|2019-10-25 07:04:57|2019-12-08 07:21:11|
| 793.5530555555556|2019-10-01 13:47:17|2019-11-03 15:20:28|
| 793.3858333333334|2019-10-01 07:21:12|2019-11-03 08:44:21|
|          793.2975|2019-10-01 13:41:00|2019-11-03 14:58:51|
| 792.9980555555555|2019-10-01 18:43:20|2019-11-03 19:43:13|
| 792.9883333333333|2019

In [7]:
# let me try ingnore the year then repeat the minus
df_fhv_hour_no_year = spark.sql("""
SELECT
    ((hour(dropOff_datetime) * 60 + minute(dropOff_datetime) + second(dropOff_datetime) / 60.0) -
    (hour(pickup_datetime) * 60 + minute(pickup_datetime) + second(pickup_datetime) / 60.0)) / 60.0 AS trip_hours,
    pickup_datetime,
    dropOff_datetime
FROM 
    fhv
WHERE
    pickup_datetime >= "2019-01-01 00:00:00"
ORDER BY
    trip_hours DESC
""")

In [8]:
df_fhv_hour_no_year.show(15)

+-------------+-------------------+-------------------+
|   trip_hours|    pickup_datetime|   dropOff_datetime|
+-------------+-------------------+-------------------+
|23.2000000000|2019-10-02 00:37:00|2019-10-02 23:49:00|
|22.8166666667|2019-10-17 00:47:00|2019-10-17 23:36:00|
|22.6000000000|2019-10-16 00:21:00|2019-10-16 22:57:00|
|22.4666666667|2019-10-15 00:49:00|2019-10-15 23:17:00|
|22.4666666667|2019-10-16 00:45:00|2019-10-16 23:13:00|
|22.2666666667|2019-10-25 00:21:00|2019-10-25 22:37:00|
|21.3333333333|2019-10-26 00:45:00|2019-10-26 22:05:00|
|21.0833333333|2019-10-18 00:16:00|2019-10-18 21:21:00|
|20.7833333333|2019-10-18 00:49:00|2019-10-18 21:36:00|
|20.6666666667|2019-10-26 00:13:00|2019-10-26 20:53:00|
|20.6333333333|2019-10-02 00:15:00|2019-10-02 20:53:00|
|20.3833333333|2019-10-17 00:33:00|2019-10-17 20:56:00|
|19.9833333333|2019-10-25 04:00:00|2019-10-25 23:59:00|
|19.9833333333|2019-10-18 04:00:00|2019-10-18 23:59:00|
|19.9833333333|2019-10-02 00:54:00|2019-10-02 20

In [44]:
#let me define the trip happens in the same day
df_fhv_hour_day = spark.sql("""
SELECT
    ((unix_timestamp(dropOff_datetime) - unix_timestamp(pickup_datetime)) / 3600) AS trip_hours,
    pickup_datetime,
    dropOff_datetime
FROM 
    fhv
WHERE
    date(pickup_datetime) = date(dropOff_datetime)
AND
    pickup_datetime >= "2019-01-01 00:00:00"
ORDER BY
    trip_hours DESC
""")

In [47]:
df_fhv_hour_day.show(15)

+------------------+-------------------+-------------------+
|        trip_hours|    pickup_datetime|   dropOff_datetime|
+------------------+-------------------+-------------------+
|              23.2|2019-10-02 00:37:00|2019-10-02 23:49:00|
|22.816666666666666|2019-10-17 00:47:00|2019-10-17 23:36:00|
|              22.6|2019-10-16 00:21:00|2019-10-16 22:57:00|
|22.466666666666665|2019-10-15 00:49:00|2019-10-15 23:17:00|
|22.466666666666665|2019-10-16 00:45:00|2019-10-16 23:13:00|
|22.266666666666666|2019-10-25 00:21:00|2019-10-25 22:37:00|
|21.333333333333332|2019-10-26 00:45:00|2019-10-26 22:05:00|
|21.083333333333332|2019-10-18 00:16:00|2019-10-18 21:21:00|
|20.783333333333335|2019-10-18 00:49:00|2019-10-18 21:36:00|
|20.666666666666668|2019-10-26 00:13:00|2019-10-26 20:53:00|
|20.633333333333333|2019-10-02 00:15:00|2019-10-02 20:53:00|
|20.383333333333333|2019-10-17 00:33:00|2019-10-17 20:56:00|
|19.983333333333334|2019-10-25 04:00:00|2019-10-25 23:59:00|
|19.983333333333334|2019

The interesting thing is the first several records have errors, but I cannot solve it, the answer will be 631,152.50 Hours

# 5. local host for Spark's User Interface is 4040

# 6. Least frequent pickup location zone
## step 1 read zone file
## step 2 join
## step 3 count then order by 

In [5]:
df_zones = spark.read.parquet("data/zones/*")

In [6]:
df_zones.show()

+----------+-------------+--------------------+------------+
|LocationID|      Borough|                Zone|service_zone|
+----------+-------------+--------------------+------------+
|         1|          EWR|      Newark Airport|         EWR|
|         2|       Queens|         Jamaica Bay|   Boro Zone|
|         3|        Bronx|Allerton/Pelham G...|   Boro Zone|
|         4|    Manhattan|       Alphabet City| Yellow Zone|
|         5|Staten Island|       Arden Heights|   Boro Zone|
|         6|Staten Island|Arrochar/Fort Wad...|   Boro Zone|
|         7|       Queens|             Astoria|   Boro Zone|
|         8|       Queens|        Astoria Park|   Boro Zone|
|         9|       Queens|          Auburndale|   Boro Zone|
|        10|       Queens|        Baisley Park|   Boro Zone|
|        11|     Brooklyn|          Bath Beach|   Boro Zone|
|        12|    Manhattan|        Battery Park| Yellow Zone|
|        13|    Manhattan|   Battery Park City| Yellow Zone|
|        14|     Brookly

In [13]:
df_join = df_fhv \
    .join(df_zones, df_fhv.PUlocationID==df_zones.LocationID, how="left") \
    .drop("dropOff_datetime", "SR_Flag", "Affiliated_base_number", "service_zone")

In [14]:
df_join.show()

+--------------------+-------------------+------------+------------+----------+---------+--------------------+
|dispatching_base_num|    pickup_datetime|PUlocationID|DOlocationID|LocationID|  Borough|                Zone|
+--------------------+-------------------+------------+------------+----------+---------+--------------------+
|              B02784|2019-10-01 09:55:38|          89|          85|        89| Brooklyn|Flatbush/Ditmas Park|
|              B02429|2019-10-21 04:15:47|         264|         264|       264|  Unknown|                  NV|
|              B01482|2019-10-19 12:00:00|         264|         264|       264|  Unknown|                  NV|
|              B03015|2019-10-11 14:28:00|         264|         216|       264|  Unknown|                  NV|
|              B01529|2019-10-21 18:00:26|         264|          80|       264|  Unknown|                  NV|
|              B00477|2019-10-03 19:30:35|         161|          14|       161|Manhattan|      Midtown Center|
|

In [16]:
df_join.registerTempTable("join_zone")

In [17]:
df_fhv_least_zone = spark.sql("""
SELECT
    Zone,
    COUNT(1) AS frequent
FROM 
    join_zone
GROUP BY
    zone
ORDER BY
    frequent
""")

In [18]:
df_fhv_least_zone.show()

+--------------------+--------+
|                Zone|frequent|
+--------------------+--------+
|         Jamaica Bay|       1|
|Governor's Island...|       2|
| Green-Wood Cemetery|       5|
|       Broad Channel|       8|
|     Highbridge Park|      14|
|        Battery Park|      15|
|Saint Michaels Ce...|      23|
|Breezy Point/Fort...|      25|
|Marine Park/Floyd...|      26|
|        Astoria Park|      29|
|    Inwood Hill Park|      39|
|       Willets Point|      47|
|Forest Park/Highl...|      53|
|  Brooklyn Navy Yard|      57|
|        Crotona Park|      62|
|        Country Club|      77|
|     Freshkills Park|      89|
|       Prospect Park|      98|
|     Columbia Street|     105|
|  South Williamsburg|     110|
+--------------------+--------+
only showing top 20 rows



The answer should be Jamaica Bay